core/store/src/flat/storage.rs

```diff
@@ -347,21 +386,71 @@ fn missing_delta_error(block_hash: &CryptoHash) -> FlatStorageError {
     FlatStorageError::StorageInternalError(format!("delta does not exist for block {block_hash}"))
 }
 
+// Determine a block to make the new flat head.
+// If `strict`, uses `block_hash` as the new flat head.
+// If not `strict`, uses the second most recent block with flat state changes from `block_hash`.
+// FlatStorageInner must be already locked.
+fn get_new_flat_head(
+    flat_storage: &FlatStorageInner,
+    block_hash: CryptoHash,
+    strict: bool,
+) -> Result<CryptoHash, FlatStorageError> {
+    if strict {
+        return Ok(block_hash);
+    }
+
+    let current_flat_head_hash = flat_storage.flat_head.hash;
+    let current_flat_head_height = flat_storage.flat_head.height;
+
+    // Keep 2 blocks with flat state changes between flat head and the final block.
+    // This is enough to make flat state snapshots function properly for state sync.
```

Comment:
super nit: for me, state sync context makes sense for general descriptions why do we care about deltas with changes, but not in inner implementation details. Maybe I'd say that FS takes care about "flat state snapshots function properly for state sync." in `update_flat_head` only.